spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/SparkParquetFileMergeRunner.java

```diff
@@ -0,0 +1,474 @@
+        group.rewrittenFiles().size(),
+        newFiles.size(),
+        groupId);
+  }
+
+  /**
+   * Distributes files evenly across the expected number of output files. This simple distribution
+   * trusts the planner's calculation of expectedOutputFiles and just divides files evenly.
+   */
+  private List<List<DataFile>> distributeFilesEvenly(Set<DataFile> files, int expectedOutputFiles) {
+    List<DataFile> fileList = Lists.newArrayList(files);
+    List<List<DataFile>> groups = Lists.newArrayList();
+
+    if (expectedOutputFiles <= 0 || fileList.isEmpty()) {
+      return groups;
+    }
+
+    int filesPerGroup = (int) Math.ceil((double) fileList.size() / expectedOutputFiles);
+
+    for (int i = 0; i < fileList.size(); i += filesPerGroup) {
+      int endIndex = Math.min(i + filesPerGroup, fileList.size());
+      groups.add(fileList.subList(i, endIndex));
+    }
+
+    return groups;
+  }
+
+  /**
+   * Performs the actual merge operation for a single task on an executor. Returns only metadata
+   * (file path and size); DataFile construction happens on the driver.
+   */
+  private static MergeResult mergeFilesForTask(MergeTaskInfo task, FileIO fileIO)
+      throws IOException {
+    // Convert file path strings to Iceberg InputFile objects
+    List<InputFile> inputFiles =
+        task.getFilePaths().stream()
+            .map(path -> fileIO.newInputFile(path))
+            .collect(Collectors.toList());
+
+    // Use OutputFileFactory to generate output file with proper naming and partition handling
+    // Encryption is handled internally based on table configuration
+    EncryptedOutputFile encryptedOutputFile =
+        task.getPartition() != null
+            ? task.getFileFactory().newOutputFile(task.getPartition())
+            : task.getFileFactory().newOutputFile();
+
+    // Merge files - schema already validated on driver, use it directly
+    boolean hasPhysicalRowIds =
+        ParquetFileMerger.mergeFilesWithOptionalRowIds(
+            inputFiles,
+            encryptedOutputFile,
+            task.getSchema(),
+            task.getFirstRowIds(),
+            task.getRowGroupSize(),
+            task.getColumnIndexTruncateLength(),
+            null);
+
+    // Get file size from the output file
+    String outputPath = encryptedOutputFile.encryptingOutputFile().location();
+    long fileSize = fileIO.newInputFile(outputPath).getLength();
+
+    // Return lightweight metadata - driver will construct DataFile with metrics
+    // firstRowId will be extracted from Parquet column statistics on the driver
+    return new MergeResult(outputPath, fileSize, task.getSpec(), task.getPartition());
+  }
+
+  /**
+   * Lightweight serializable task containing only the essential information needed for merging.
+   * Uses simple types ({@code String}, {@code List<String>}) that serialize reliably.
+   */
+  private static class MergeTaskInfo implements Serializable {
+    private final String taskId;
```

Comment:
Is this used?

Reply:
Good catch! The taskId was only used for logging on the driver before serializing the task. The accessor method was never called. I've removed it from MergeTaskInfo to reduce serialization overhead and eliminate the unused field.